File: frontend/plugin-manager/src/api/pluginCli.ts
/**
 * neko-plugin-cli 相关 API
 */
import { get, post } from './index'
import { API_BASE_URL } from '@/utils/constants'

export type PluginCliConflictStrategy = 'rename' | 'fail'
export type PluginCliBuildMode = 'selected' | 'single' | 'bundle' | 'all'

export interface PluginCliPluginRef {
  root_id: 'builtin' | 'user'
  directory_name: string
  plugin_id?: string
  label?: string
}

export interface PluginCliBuildRequest {
  mode: PluginCliBuildMode
  plugin?: string
  plugins?: string[]
  plugin_ref?: PluginCliPluginRef
  plugin_refs?: PluginCliPluginRef[]
  out?: string
  target_dir?: string
  keep_staging?: boolean
  bundle_id?: string
  package_name?: string
  package_description?: string
  version?: string
}

export interface PluginCliPluginItem {
  plugin: string
  error: string
}

export interface PluginCliBuildResult {
  plugin_id: string
  package_type: string
  plugin_ids: string[]
  package_name?: string
  version?: string
  package_path: string
  staging_dir?: string | null
  profile_files: string[]
  staged_files: string[]
  payload_hash: string
  package_size_bytes: number
  staged_file_count: number
  profile_file_count: number
}

export interface PluginCliBuildResponse {
  built: PluginCliBuildResult[]
  built_count: number
  failed: PluginCliPluginItem[]
  failed_count: number
  ok: boolean
}

export interface PluginCliPackageRef {
  package: string
}

export interface PluginCliInspectedPlugin {
  plugin_id: string
  archive_path: string
  has_plugin_toml: boolean
}

export interface PluginCliInspectResponse {
  package_path: string
  package_type: string
  package_id: string
  schema_version: string
  package_name: string
  package_description: string
  version: string
  metadata_found: boolean
  payload_hash: string
  payload_hash_verified: boolean | null
  plugins: PluginCliInspectedPlugin[]
  profile_names: string[]
  plugin_count: number
  profile_count: number
}

export interface PluginCliVerifyResponse extends PluginCliInspectResponse {
  ok: boolean
}

export interface PluginCliInstallRequest {
  package: string
  plugins_root?: string
  profiles_root?: string
  on_conflict?: PluginCliConflictStrategy
}

export interface PluginCliInstalledPlugin {
  source_folder: string
  target_plugin_id: string
  target_dir: string
  renamed: boolean
}

export interface PluginCliInstallResponse {
  package_path: string
  package_type: string
  package_id: string
  plugins_root: string
  profiles_root?: string | null
  installed_plugins: PluginCliInstalledPlugin[]
  profile_dir?: string | null
  metadata_found: boolean
  payload_hash: string
  payload_hash_verified: boolean | null
  conflict_strategy: PluginCliConflictStrategy
  installed_plugin_count: number
}

export interface PluginCliAnalyzeRequest {
  plugins?: string[]
  plugin_refs?: PluginCliPluginRef[]
  current_sdk_version?: string
}

export interface PluginCliSharedDependency {
  name: string
  plugin_ids: string[]
  requirement_texts: Record<string, string>
  plugin_count: number
}

export interface PluginCliBundleSdkAnalysis {
  kind: string
  plugin_specifiers: Record<string, string>
  has_overlap: boolean
  matching_versions: string[]
  current_sdk_version: string
  current_sdk_supported_by_all: boolean | null
}

export interface PluginCliAnalyzeResponse {
  plugin_ids: string[]
  shared_dependencies: PluginCliSharedDependency[]
  common_dependencies: PluginCliSharedDependency[]
  sdk_supported_analysis?: PluginCliBundleSdkAnalysis | null
  sdk_recommended_analysis?: PluginCliBundleSdkAnalysis | null
  plugin_count: number
}

export interface PluginCliLocalPluginsResponse {
  plugins: string[]
  plugin_refs?: PluginCliPluginRef[]
  count: number
}

export interface PluginCliLocalPackageItem {
  name: string
  path: string
  suffix: string
  size_bytes: number
  modified_at: string
}

export interface PluginCliLocalPackagesResponse {
  packages: PluginCliLocalPackageItem[]
  count: number
  target_dir: string
}

/**
 * 列出当前本地可构建插件
 */
export function getPluginCliPlugins(): Promise<PluginCliLocalPluginsResponse> {
  return get('/plugin-cli/plugins')
}

/**
 * 列出当前 target 目录中的本地包
 */
export function getPluginCliPackages(): Promise<PluginCliLocalPackagesResponse> {
  return get('/plugin-cli/packages')
}

/**
 * 构建一个或多个插件
 */
export function buildPluginCli(payload: PluginCliBuildRequest): Promise<PluginCliBuildResponse> {
  return post('/plugin-cli/build', payload)
}

/**
 * 检查包内容
 */
export function inspectPluginPackage(payload: PluginCliPackageRef): Promise<PluginCliInspectResponse> {
  return post('/plugin-cli/inspect', payload)
}

/**
 * 校验包的 payload hash
 */
export function verifyPluginPackage(payload: PluginCliPackageRef): Promise<PluginCliVerifyResponse> {
  return post('/plugin-cli/verify', payload)
}

/**
 * 安装插件包或整合包
 */
export function installPluginPackage(payload: PluginCliInstallRequest): Promise<PluginCliInstallResponse> {
  return post('/plugin-cli/install', payload)
}

/**
 * 分析多个插件的整合包兼容性
 */
export function analyzePluginBundle(payload: PluginCliAnalyzeRequest): Promise<PluginCliAnalyzeResponse> {
  return post('/plugin-cli/analyze', payload)
}

// ── Upload & Download ─────────────────────────────────────────────────

export interface PluginCliUploadResult {
  name: string
  path: string
  size_bytes: number
  modified_at: string
}

export interface PluginCliUploadAndInstallResult {
  upload: PluginCliUploadResult
  install: PluginCliInstallResponse
}

/**
 * 上传插件包文件到服务器
 */
export function uploadPluginPackage(file: File): Promise<PluginCliUploadResult> {
  const formData = new FormData()
  formData.append('file', file)
  return post('/plugin-cli/upload', formData, {
    timeout: 120_000,
  })
}

/**
 * 上传插件包并立即安装
 */
export function uploadAndInstallPlugin(
  file: File,
  options?: { onConflict?: PluginCliConflictStrategy },
): Promise<PluginCliUploadAndInstallResult> {
  const formData = new FormData()
  formData.append('file', file)
  const params = new URLSearchParams()
  if (options?.onConflict) {
    params.set('on_conflict', options.onConflict)
  }
  const query = params.toString()
  const url = `/plugin-cli/upload-and-install${query ? `?${query}` : ''}`
  return post(url, formData, {
    timeout: 120_000,
  })
}

/**
 * 构建插件包下载 URL（用于浏览器直接下载）
 */
export function getPluginPackageDownloadUrl(packagePath: string): string {
  const params = new URLSearchParams({ package: packagePath })
  return `${API_BASE_URL}/plugin-cli/download?${params.toString()}`
}

/**
 * 触发浏览器下载插件包
 */
export function downloadPluginPackage(packagePath: string): void {
  const url = getPluginPackageDownloadUrl(packagePath)
  // Extract just the filename for the browser download hint
  const filename = packagePath.split('/').pop() || packagePath.split('\\').pop() || packagePath
  const link = document.createElement('a')
  link.href = url
  link.download = filename
  link.style.display = 'none'
  document.body.appendChild(link)
  link.click()
  document.body.removeChild(link)
}

File: plugin/server/routes/plugin_cli.py
from __future__ import annotations

from fastapi import APIRouter, File, HTTPException, Query, UploadFile
from fastapi.responses import FileResponse
from pydantic import BaseModel, Field, model_validator

from plugin.logging_config import get_logger
from plugin.server.application.plugin_cli import PluginCliService
from plugin.server.domain.errors import ServerDomainError
from plugin.server.infrastructure.auth import require_admin
from plugin.server.infrastructure.error_mapping import raise_http_from_domain

router = APIRouter()
logger = get_logger("server.routes.plugin_cli")
service = PluginCliService()


class PluginCliPluginRef(BaseModel):
    root_id: str = Field(pattern="^(builtin|user)$")
    directory_name: str


class PluginCliPluginRefResponse(PluginCliPluginRef):
    plugin_id: str = ""
    label: str = ""


class PluginCliBuildRequest(BaseModel):
    mode: str = Field(default="selected", pattern="^(selected|single|bundle|all)$")
    plugin: str | None = None
    plugins: list[str] = Field(default_factory=list)
    plugin_ref: PluginCliPluginRef | None = None
    plugin_refs: list[PluginCliPluginRef] = Field(default_factory=list)
    out: str | None = None
    target_dir: str | None = None
    keep_staging: bool = False
    bundle_id: str | None = None
    package_name: str | None = None
    package_description: str | None = None
    version: str | None = None

    @model_validator(mode="after")
    def _validate_mode_payload(self) -> "PluginCliBuildRequest":
        if self.mode == "single" and not (self.plugin_ref or self.plugin):
            raise ValueError("plugin_ref or plugin is required when mode=single")
        if self.mode in {"selected", "bundle"} and not (self.plugin_refs or self.plugins):
            raise ValueError("plugin_refs or plugins is required when mode=selected or mode=bundle")
        return self


class PluginCliPackageRequest(BaseModel):
    package: str


class PluginCliInstallRequest(BaseModel):
    package: str
    plugins_root: str | None = None
    profiles_root: str | None = None
    on_conflict: str = Field(default="rename", pattern="^(rename|fail)$")


class PluginCliAnalyzeRequest(BaseModel):
    plugins: list[str] = Field(default_factory=list)
    plugin_refs: list[PluginCliPluginRef] = Field(default_factory=list)
    current_sdk_version: str | None = None


class PluginCliPluginListResponse(BaseModel):
    plugins: list[str]
    plugin_refs: list[PluginCliPluginRefResponse] = Field(default_factory=list)
    count: int


class PluginCliLocalPackageItem(BaseModel):
    name: str
    path: str
    suffix: str
    size_bytes: int
    modified_at: str


class PluginCliPackageListResponse(BaseModel):
    packages: list[PluginCliLocalPackageItem]
    count: int
    target_dir: str


class PluginCliBuildFailure(BaseModel):
    plugin: str
    error: str


class PluginCliBuildResultResponse(BaseModel):
    plugin_id: str
    package_type: str
    plugin_ids: list[str]
    package_name: str = ""
    version: str = ""
    package_path: str
    staging_dir: str | None = None
    profile_files: list[str]
    staged_files: list[str]
    payload_hash: str
    package_size_bytes: int
    staged_file_count: int
    profile_file_count: int
    plugin_count: int


class PluginCliBuildResponse(BaseModel):
    built: list[PluginCliBuildResultResponse]
    built_count: int
    failed: list[PluginCliBuildFailure]
    failed_count: int
    ok: bool


class PluginCliInspectedPluginResponse(BaseModel):
    plugin_id: str
    archive_path: str
    has_plugin_toml: bool


class PluginCliDependencyPluginResponse(BaseModel):
    plugin_id: str
    python_requirements: list[str]
    host_python_requirements: list[str]
    plugin_dependencies: list[str]
    advanced_plugin_dependencies: list[dict[str, object]]
    vendor_path: str = ""
    vendor_present: bool = False


class PluginCliDependencySummaryResponse(BaseModel):
    schema_version: str = ""
    plugins: list[PluginCliDependencyPluginResponse]
    plugin_count: int


class PluginCliInspectResponse(BaseModel):
    package_path: str
    package_type: str
    package_id: str
    schema_version: str = ""
    package_name: str = ""
    package_description: str = ""
    version: str = ""
    metadata_found: bool
    payload_hash: str = ""
    payload_hash_verified: bool | None = None
    plugins: list[PluginCliInspectedPluginResponse]
    profile_names: list[str]
    plugin_count: int
    profile_count: int
    dependencies: PluginCliDependencySummaryResponse | None = None


class PluginCliVerifyResponse(PluginCliInspectResponse):
    ok: bool


class PluginCliInstalledPluginResponse(BaseModel):
    source_folder: str
    target_plugin_id: str
    target_dir: str
    renamed: bool


class PluginCliInstallResponse(BaseModel):
    package_path: str
    package_type: str
    package_id: str
    plugins_root: str
    profiles_root: str | None = None
    installed_plugins: list[PluginCliInstalledPluginResponse]
    profile_dir: str | None = None
    metadata_found: bool
    payload_hash: str = ""
    payload_hash_verified: bool | None = None
    conflict_strategy: str
    installed_plugin_count: int


class PluginCliSharedDependencyResponse(BaseModel):
    name: str
    plugin_ids: list[str]
    requirement_texts: dict[str, str]
    plugin_count: int


class PluginCliBundleSdkAnalysisResponse(BaseModel):
    kind: str
    plugin_specifiers: dict[str, str]
    has_overlap: bool
    matching_versions: list[str]
    current_sdk_version: str = ""
    current_sdk_supported_by_all: bool | None = None


class PluginCliAnalyzeResponse(BaseModel):
    plugin_ids: list[str]
    shared_dependencies: list[PluginCliSharedDependencyResponse]
    common_dependencies: list[PluginCliSharedDependencyResponse]
    sdk_supported_analysis: PluginCliBundleSdkAnalysisResponse | None = None
    sdk_recommended_analysis: PluginCliBundleSdkAnalysisResponse | None = None
    plugin_count: int


class PluginCliUploadResponse(BaseModel):
    name: str
    path: str
    size_bytes: int
    modified_at: str


class PluginCliUploadAndInstallResponse(BaseModel):
    upload: PluginCliUploadResponse
    install: PluginCliInstallResponse
    install_source_warning: str | None = None


@router.get("/plugin-cli/plugins", response_model=PluginCliPluginListResponse)
async def list_plugin_cli_plugins(_: str = require_admin) -> dict[str, object]:
    try:
        return await service.list_local_plugins()
    except ServerDomainError as error:
        raise_http_from_domain(error, logger=logger)


@router.get("/plugin-cli/packages", response_model=PluginCliPackageListResponse)
async def list_plugin_cli_packages(_: str = require_admin) -> dict[str, object]:
    try:
        return await service.list_local_packages()
    except ServerDomainError as error:
        raise_http_from_domain(error, logger=logger)


@router.post("/plugin-cli/build", response_model=PluginCliBuildResponse)
async def plugin_cli_build(
    payload: PluginCliBuildRequest,
    _: str = require_admin,
) -> dict[str, object]:
    try:
        return await service.build(
            mode=payload.mode,
            plugin=payload.plugin,
            plugins=payload.plugins,
            plugin_ref=payload.plugin_ref.model_dump() if payload.plugin_ref else None,
            plugin_refs=[item.model_dump() for item in payload.plugin_refs],
            out=payload.out,
            target_dir=payload.target_dir,
            keep_staging=payload.keep_staging,
            bundle_id=payload.bundle_id,
            package_name=payload.package_name,
            package_description=payload.package_description,
            version=payload.version,
        )
    except ServerDomainError as error:
        raise_http_from_domain(error, logger=logger)


@router.post("/plugin-cli/inspect", response_model=PluginCliInspectResponse)
async def plugin_cli_inspect(
    payload: PluginCliPackageRequest,
    _: str = require_admin,
) -> dict[str, object]:
    try:
        return await service.inspect(package=payload.package)
    except ServerDomainError as error:
        raise_http_from_domain(error, logger=logger)


@router.post("/plugin-cli/verify", response_model=PluginCliVerifyResponse)
async def plugin_cli_verify(
    payload: PluginCliPackageRequest,
    _: str = require_admin,
) -> dict[str, object]:
    try:
        return await service.verify(package=payload.package)
    except ServerDomainError as error:
        raise_http_from_domain(error, logger=logger)


@router.post("/plugin-cli/install", response_model=PluginCliInstallResponse)
async def plugin_cli_install(
    payload: PluginCliInstallRequest,
    _: str = require_admin,
) -> dict[str, object]:
    try:
        return await service.install(
            package=payload.package,
            plugins_root=payload.plugins_root,
            profiles_root=payload.profiles_root,
            on_conflict=payload.on_conflict,
        )
    except ServerDomainError as error:
        raise_http_from_domain(error, logger=logger)


@router.post("/plugin-cli/analyze", response_model=PluginCliAnalyzeResponse)
async def plugin_cli_analyze(
    payload: PluginCliAnalyzeRequest,
    _: str = require_admin,
) -> dict[str, object]:
    try:
        return await service.analyze(
            plugins=payload.plugins,
            plugin_refs=[item.model_dump() for item in payload.plugin_refs],
            current_sdk_version=payload.current_sdk_version,
        )
    except ServerDomainError as error:
        raise_http_from_domain(error, logger=logger)


# ── Upload & Download ──────────────────────────────────────────────────


@router.post("/plugin-cli/upload", response_model=PluginCliUploadResponse)
async def plugin_cli_upload(
    file: UploadFile = File(...),
    _: str = require_admin,
) -> dict[str, object]:
    """Upload a plugin package file (.neko-plugin / .neko-bundle) to the server.

    The file is saved to the packages target directory and can subsequently be
    passed to ``/plugin-cli/install`` or ``/plugin-cli/inspect``.
    """
    try:
        content = await file.read()
        return await service.save_uploaded_package(
            filename=file.filename or "unknown.neko-plugin",
            content=content,
        )
    except ServerDomainError as error:
        raise_http_from_domain(error, logger=logger)
    except Exception:
        logger.exception("Unexpected error during plugin package upload")
        raise HTTPException(status_code=500, detail="Internal server error during upload")


@router.post("/plugin-cli/upload-and-install", response_model=PluginCliUploadAndInstallResponse)
async def plugin_cli_upload_and_install(
    file: UploadFile = File(...),
    on_conflict: str = Query(default="rename", pattern="^(rename|fail)$"),
    _: str = require_admin,
) -> dict[str, object]:
    """Upload a plugin package and immediately install it.

    Combines upload + install into a single request for convenience.
    """
    try:
        content = await file.read()
        return await service.upload_and_install(
            filename=file.filename or "unknown.neko-plugin",
            content=content,
            on_conflict=on_conflict,
        )
    except ServerDomainError as error:
        raise_http_from_domain(error, logger=logger)
    except Exception:
        logger.exception("Unexpected error during plugin package upload-and-install")
        raise HTTPException(status_code=500, detail="Internal server error during upload-and-install")


@router.get("/plugin-cli/download")
async def plugin_cli_download(
    package: str = Query(..., description="Package filename or path within the target directory"),
    _: str = require_admin,
) -> FileResponse:
    """Download a plugin package file from the server."""
    try:
        resolved = service.resolve_download_path(package)
        return FileResponse(
            str(resolved),
            filename=resolved.name,
            media_type="application/octet-stream",
        )
    except ServerDomainError as error:
        raise_http_from_domain(error, logger=logger)


# ── Legacy route aliases (backward compatibility with existing frontend) ──


@router.post("/plugin-cli/pack", include_in_schema=False)
async def plugin_cli_pack_legacy(
    payload: PluginCliBuildRequest,
    _: str = require_admin,
) -> dict[str, object]:
    """Legacy alias for /plugin-cli/build. Translates response keys."""
    result = await plugin_cli_build(payload, _)
    # Translate new keys to legacy keys expected by frontend
    if isinstance(result, dict):
        translated = dict(result)
        if "built" in translated:
            translated["packed"] = translated.pop("built")
        if "built_count" in translated:
            translated["packed_count"] = translated.pop("built_count")
        return translated
    return result


@router.post("/plugin-cli/unpack", include_in_schema=False)
async def plugin_cli_unpack_legacy(
    payload: PluginCliInstallRequest,
    _: str = require_admin,
) -> dict[str, object]:
    """Legacy alias for /plugin-cli/install. Translates response keys."""
    result = await plugin_cli_install(payload, _)
    # Translate new keys to legacy keys expected by frontend
    if isinstance(result, dict):
        translated = dict(result)
        if "installed_plugins" in translated:
            translated["unpacked_plugins"] = translated.pop("installed_plugins")
        if "installed_plugin_count" in translated:
            translated["unpacked_plugin_count"] = translated.pop("installed_plugin_count")
        return translated
    return result


@router.post("/plugin-cli/upload-and-unpack", include_in_schema=False)
async def plugin_cli_upload_and_unpack_legacy(
    file: UploadFile = File(...),
    on_conflict: str = Query(default="rename", pattern="^(rename|fail)$"),
    _: str = require_admin,
) -> dict[str, object]:
    """Legacy alias for /plugin-cli/upload-and-install. Translates response keys."""
    result = await plugin_cli_upload_and_install(file, on_conflict=on_conflict, _=_)
    # Translate nested install keys
    if isinstance(result, dict) and isinstance(result.get("install"), dict):
        install = dict(result["install"])
        if "installed_plugins" in install:
            install["unpacked_plugins"] = install.pop("installed_plugins")
        if "installed_plugin_count" in install:
            install["unpacked_plugin_count"] = install.pop("installed_plugin_count")
        result = {key: value for key, value in result.items() if key != "install"}
        result["unpack"] = install
    return result
